Add <input type=checkbox switch> macOS implementation

annevk opened the pull request:
#### 97661e2a2d6838b9b044ac95164e8e20980c618a
<pre>
Add &lt;input type=checkbox switch&gt; macOS implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=263372">https://bugs.webkit.org/show_bug.cgi?id=263372</a>
rdar://117199685

Reviewed by NOBODY (OOPS!).

Add a shadow root for &lt;input type=checkbox switch&gt; in a matter
consistent with the CSS WG plan for a track and thumb pseudo-element.

Add a macOS implementation that has these known shortcomings:

- There&apos;s no animation support.
- There&apos;s no support for dragging the thumb.
- Toggling &quot;Increase contrast&quot; or &quot;Differentiate without color&quot; doesn&apos;t
  update the look of the control dynamically.

This builds on excellent prototyping work done by Lily Spiniolas.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/spi/mac/CoreUISPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSwitchSPI.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/css/html.css:
(input[type=&quot;checkbox&quot;][switch]):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::createShadowSubtree):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::needsShadowSubtree const):
(WebCore::CheckboxInputType::createShadowSubtree):
(WebCore::CheckboxInputType::attributeChanged):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::createShadowSubtree):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::createShadowSubtree):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::destroyShadowSubtree):
* Source/WebCore/html/InputType.h:
(WebCore::InputType::needsShadowSubtree const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::createShadowSubtree):
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::createShadowSubtree):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::createShadowSubtree):
* Source/WebCore/html/shadow/SwitchThumbElement.cpp: Added.
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::SwitchThumbElement):
(WebCore::SwitchThumbElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchThumbElement.h: Added.
* Source/WebCore/html/shadow/SwitchTrackElement.cpp: Added.
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::SwitchTrackElement):
(WebCore::SwitchTrackElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchTrackElement.h: Added.
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h: Added.
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h: Added.
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.h:
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.mm:
(WebCore::ControlFactoryIOS::createPlatformSwitchThumb):
(WebCore::ControlFactoryIOS::createPlatformSwitchTrack):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::switchControl const):
(WebCore::ControlFactoryMac::createPlatformSwitchThumb):
(WebCore::ControlFactoryMac::createPlatformSwitchTrack):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm: Added.
(WebCore::SwitchThumbMac::SwitchThumbMac):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm: Added.
(WebCore::SwitchTrackMac::SwitchTrackMac):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isPressed const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97661e2a2d6838b9b044ac95164e8e20980c618a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22929 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/960 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24840 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21230 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23465 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22115 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23170 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/465 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19885 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25697 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/23115 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/375 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26970 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19961 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21031 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24829 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22292 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/464 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18273 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/29715 "Built successfully") | 
| [❌ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/22754 "Failed resultsdbpy unit tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/379 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20556 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6152 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/875 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/28113 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/631 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6016 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/shadow/SwitchThumbElement.cpp:
```diff
@@ -0,0 +1,65 @@
+    auto elementStyle = resolveStyle(resolutionContext);
+    if (hostStyle->effectiveAppearance() == StyleAppearance::Auto) {
+        elementStyle.style->setEffectiveAppearance(StyleAppearance::SwitchThumb);
+        elementStyle.style->setWidth({ 38, LengthType::Fixed });
+        elementStyle.style->setHeight({ 24, LengthType::Fixed });
```
It would be good to have confirmation of these sizes.

nt1m replied:
Those width/height/margin adjustments should probably be done by RenderTheme when it comes across a StyleAppearance::SwitchThumb

pxlcoder replied:
To be even more specific, adjustments done in `RenderThemeMac` overriding a method in `RenderTheme` since these numbers could be platform specific.

A single fixed size may also be incorrect, because AppKit offers 4 fixed sizes. See the use of `NSControlSize` in our codebase.

annevk commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
```diff
@@ -81,6 +83,7 @@ class ControlFactoryMac final : public ControlFactoryCocoa {
     NSMenu *searchMenuTemplate() const;
     NSSliderCell *sliderCell() const;
     NSTextFieldCell *textFieldCell() const;
+    WebSwitchView *webSwitchView() const;
```
Looking at the code again this should just hold an NSSwitch directly. Should it be called `switchTemplate()` similar to the `NSMenu` one above?

pxlcoder replied:
Yes, let's just use `NSSwitch` rather than a wrapper. The name for the `NSMenu` variable is derived from the fact that it's used with `setSearchMenuTemplate:`.

I'm okay with `switchControl()` or `switchView()`. Since `NSSwitch` is a subclass of `NSControl`, which is a subclass of `NSView`.

annevk commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -0,0 +1,112 @@
+    auto inflatedRect = rectForBounds(borderRect.rect(), style);
+
+    // FIXME: _trackFrameInRect relies on interface definition in SwitchThumbMac
+    NSRect trackRect = [[m_switchPtr nativeSwitch] _trackFrameInRect:inflatedRect];
```
I don't know how to improve this.

pxlcoder replied:
The shared header I mentioned above is the correct thing to do here.

annevk reviewed (commented):
@pxlcoder I'd very much appreciate it if you could review this. I highlighted three parts I consider somewhat problematic.

nt1m commented on Source/WebCore/html/shadow/SwitchTrackElement.cpp:
```diff
@@ -0,0 +1,64 @@
+    auto elementStyle = resolveStyle(resolutionContext);
+    if (hostStyle->effectiveAppearance() == StyleAppearance::Auto) {
+        elementStyle.style->setEffectiveAppearance(StyleAppearance::SwitchTrack);
+        elementStyle.style->setWidth({ 38, LengthType::Fixed });
+        elementStyle.style->setHeight({ 24, LengthType::Fixed });
```
ditto

pxlcoder commented on Source/WebCore/css/html.css:
```diff
@@ -844,6 +844,10 @@ input:is([type="radio"], [type="checkbox"]) {
 #endif
 }
 
+input[type="checkbox"][switch] {
+    margin: initial;
+}
```
For correctness, we should probably put this in a separate stylesheet, which is only loaded if the feature flag is enabled. See `UserAgentStyle::ensureDefaultStyleSheetsForElement`.

I would recommend doing that in a separate change to avoid further increasing the size of this one.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -58,10 +63,32 @@ String CheckboxInputType::valueMissingText() const
+    ASSERT(needsShadowSubtree());
+    ASSERT(element()->userAgentShadowRoot());
+
+    Document& document = element()->document();
```
`Ref document = ...` – we should be using smart pointer where possible in new code.

pxlcoder commented on Source/WebCore/html/BaseDateAndTimeInputType.cpp:
```diff
@@ -327,8 +327,8 @@ void BaseDateAndTimeInputType::showPicker()
 void BaseDateAndTimeInputType::createShadowSubtree()
 {
-    ASSERT(needsShadowSubtree());
     ASSERT(element());
+    ASSERT(needsShadowSubtree());
```
Can you explain the reasoning behind this reordering?

pxlcoder commented on Source/WebCore/html/InputType.cpp:
```diff
@@ -677,6 +677,7 @@ void InputType::destroyShadowSubtree()
         return;
 
     root->removeChildren();
+    m_hasCreatedShadowSubtree = false;
```
I'm curious why this is needed now, but not before?

pxlcoder commented on Source/WebCore/html/InputType.h:
```diff
@@ -208,7 +208,7 @@ class InputType : public RefCounted<InputType> {
     bool isLabelable() const;
     bool isEnumeratable() const;
-    bool needsShadowSubtree() const { return !nonShadowRootTypes.contains(m_type); }
+    virtual bool needsShadowSubtree() const { return !nonShadowRootTypes.contains(m_type); }
```
There are performance considerations around shadow subtree creation. I wonder if it's better or worse to:
- Introduce a new virtual method
- Always create a shadow subtree for checkboxes

cc @cdumez @heycam

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -0,0 +1,107 @@
+- (NSRect)_trackFrameInRect:(NSRect)boundsRect;
+- (NSRect)_knobFrameInTrackFrame:(NSRect)trackFrameRect thumbIsOnRight:(BOOL)isOnRight;
+
+@end
```
This should be moved to an `NSSwitchSPI.h` file, like the others in `Source/WebCore/PAL/pal/spi/mac/`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -0,0 +1,112 @@
+{
+    ControlStyle cs(style);
+    auto rect = bounds;
+    rect.inflateY(-1/ 2);
```
`-1 / 2.f`

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -0,0 +1,107 @@
+{
+    ControlStyle cs(style);
+    auto rect = bounds;
+    rect.inflateY(-1/ 2);
```
`-1 / 2.f`

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -0,0 +1,107 @@
+
+FloatRect SwitchThumbMac::rectForBounds(const FloatRect& bounds, const ControlStyle& style) const
+{
+    ControlStyle cs(style);
```
This appears to be unused.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -0,0 +1,112 @@
+
+FloatRect SwitchTrackMac::rectForBounds(const FloatRect& bounds, const ControlStyle& style) const
+{
+    ControlStyle cs(style);
```
This appears to be unused.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -0,0 +1,112 @@
+    GraphicsContextStateSaver stateSaver(context);
+
+    [[NSAppearance currentDrawingAppearance] _drawInRect: NSMakeRect(0, 0, NSWidth(trackRect), NSHeight(trackRect)) context:cgContextForMask options:@{
+        (__bridge NSString *)kCUIWidgetKey: (__bridge NSString *)(kCUIWidgetSwitchFillMask),
```
Don't need parentheses around `kCUIWidgetSwitchFillMask`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:
```diff
@@ -176,4 +176,16 @@ - (CFDictionaryRef)_coreUIDrawOptionsWithFrame:(NSRect)cellFrame inView:(NSView
+{
+    _nativeSwitch = adoptNS([[NSSwitch alloc] init]);
+    [_nativeSwitch setWantsLayer:YES];
+    [_nativeSwitch setControlSize:NSControlSizeRegular];
```
We should support all 4  `NSControlSizes` similar to checkboxes and radio buttons.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1284,7 +1288,8 @@ bool RenderTheme::isWindowActive(const RenderObject& renderer) const
 {
-    return is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).shouldAppearChecked();
+    return (is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).shouldAppearChecked())
+        || (o.node()->shadowHost() && is<HTMLInputElement>(o.node()->shadowHost()) && downcast<HTMLInputElement>(*o.node()->shadowHost()).shouldAppearChecked());
```
This would be more readable as two separate statements, the use of `dynamicDowncast`, and an early return.

pxlcoder reviewed (commented):
This is a great start!

Some larger comments:
1. I think we should to split up the DOM/element changes and the painting into two separate PRs.
2. Currently, the thumb is being sized to fit the entire track, and we're simply changing the painting inside the rect, in order to represent on/off. Instead, we should match the `<input type=range>` implementation, and move the thumb element itself. That doesn't necessarily have to be done in the same PR as the macOS painting work though – it can be done as a follow-up.

annevk committed:
Add <input type=checkbox switch> macOS implementation
https://bugs.webkit.org/show_bug.cgi?id=263372
rdar://117199685

Reviewed by NOBODY (OOPS!).

Add a shadow root for <input type=checkbox switch> in a matter
consistent with the CSS WG plan for a track and thumb pseudo-element.

Add a macOS implementation that has these known shortcomings:

- There's no animation support.
- There's no support for dragging the thumb.
- Toggling "Increase contrast" or "Differentiate without color" doesn't
  update the look of the control dynamically.

This builds on excellent prototyping work done by Lily Spiniolas.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/spi/mac/CoreUISPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSwitchSPI.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/css/html.css:
(input[type="checkbox"][switch]):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::createShadowSubtree):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::needsShadowSubtree const):
(WebCore::CheckboxInputType::createShadowSubtree):
(WebCore::CheckboxInputType::attributeChanged):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::createShadowSubtree):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::createShadowSubtree):
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::destroyShadowSubtree):
* Source/WebCore/html/InputType.h:
(WebCore::InputType::needsShadowSubtree const):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::createShadowSubtree):
* Source/WebCore/html/SearchInputType.cpp:
(WebCore::SearchInputType::createShadowSubtree):
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::createShadowSubtree):
* Source/WebCore/html/shadow/SwitchThumbElement.cpp: Added.
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::SwitchThumbElement):
(WebCore::SwitchThumbElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchThumbElement.h: Added.
* Source/WebCore/html/shadow/SwitchTrackElement.cpp: Added.
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::SwitchTrackElement):
(WebCore::SwitchTrackElement::resolveCustomStyle):
* Source/WebCore/html/shadow/SwitchTrackElement.h: Added.
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h: Added.
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h: Added.
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.h:
* Source/WebCore/platform/graphics/ios/controls/ControlFactoryIOS.mm:
(WebCore::ControlFactoryIOS::createPlatformSwitchThumb):
(WebCore::ControlFactoryIOS::createPlatformSwitchTrack):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::switchControl const):
(WebCore::ControlFactoryMac::createPlatformSwitchThumb):
(WebCore::ControlFactoryMac::createPlatformSwitchTrack):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm: Added.
(WebCore::SwitchThumbMac::SwitchThumbMac):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm: Added.
(WebCore::SwitchTrackMac::SwitchTrackMac):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isPressed const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator<<):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
